refactor(ivy): LView is a proper linked list

benlesh opened the pull request:
- TView no longer stores childIndex
- LView now as CHILD_HEAD and CHILD_TAIL

TView used to store the head of the list, therefor all LViews had to have the same head, which is incorrect.

benlesh committed:
refactor(ivy): LView is a proper linked list

- TView no longer stores childIndex
- LView now as CHILD_HEAD and CHILD_TAIL

TView used to store the head of the list, therefor all LViews had to have the same head, which is incorrect.

kara reviewed (changes requested):
Looks good, but needs a test for the new behavior we want to support (e.g. using injector.get to get `ViewContainerRef`.

Also needs symbol test updates and linting.

benlesh committed:
refactor(ivy): LContainer now stored in LView[PARENT]

- Removes CONTAINER_INDEX
- LView[PARENT] now contains LContainer when necessary
- Removes now unused arguments to methods after refactor

gkalpak reviewed (commented):
Isn't it possible to also remove views? Shouldn't `CHILD_HEAD`/`CHILD_TAIL` be updatd in that case?

benlesh commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -234,18 +235,19 @@ export function addRemoveViewFromContainer(
     return cleanUpView(rootView);
   }
-  let viewOrContainer: LView|LContainer|null = getLViewChild(rootView);
+  let viewOrContainer: LView|LContainer|null = rootView[CHILD_HEAD];
```
Hey, @gkalpak ... I guess I'm not sure what you mean (and the code has shifted a little since this comment), can you elaborate?

gkalpak replied:
I mean you already have `rootChildHead = rootView[CHILD_HEAD]` a couple of lines above, so maybe you could reuse the same var. E.g.:

```diff
-let viewOrContainer: LView|LContainer|null = rootView[CHILD_HEAD];
+let viewOrContainer: LView|LContainer|null = rootChildHead;
```

...or even...

```diff
-const rootChildHead = rootView[CHILD_HEAD];
-if (!rootChildHead) {
+let viewOrContainer = rootView[CHILD_HEAD];
+if (!viewOrContainer) {
   return cleanUpView(rootView);
 }
-let viewOrContainer: LView|LContainer|null = rootView[CHILD_HEAD];
```

kara replied:
+1 to George's comment

AndrewKushnir commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2056,8 +2054,10 @@ function generateInitialInputs(
+    hostNative: RElement | RComment | StylingContext | LView, currentView: LView, native: RComment,
     isForViewContainerRef?: boolean): LContainer {
+  ngDevMode && assertDomNode(native);
+  ngDevMode && assertLView(currentView, true);
```
I think it might be a bit hard to follow what `true` stands here for.
As an option, can we have 2 asserts here, like:
```
ngDevMode &&
    assertDefined(currentView, 'Current view must be defined') &&
    assertLView(currentView);
```
and keep just `assertLView` where required is false (default).
@benlesh what do you think?

benlesh replied:
I sort of see what you mean. But that also makes the assertions a lot "busier" when reading the code. I _think_ I'd prefer to keep the assertions a little more terse, myself, even though I'm generally in favor of more explicit, readable code. @mhevery what do you think?

mhevery replied:
I am fine either way.

benlesh replied:
Can we do it in another PR? The FW-869 work I'm doing is built on top of this.

AndrewKushnir replied:
Sure, no problem, thanks the reply.

benlesh commented:
presubmit: https://test.corp.google.com/ui#id=OCL:232328029:BASE:232328145:1549306899223:9eaec735

kara commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2500,7 +2502,7 @@ export function projection(nodeIndex: number, selectorIndex: number = 0, attrs?:
   const componentNode = componentView[HOST_NODE] as TElementNode;
   let nodeToProject = (componentNode.projection as(TNode | null)[])[selectorIndex];
-  let projectedView = componentView[PARENT] !;
+  let projectedView = getLViewParent(componentView) !;
```
No need to wrap this one - by definition, the component view is the view just below the component boundary, so it can't have a container on top.

kara commented on packages/core/src/render3/instructions.ts:
```diff
@@ -2551,20 +2553,20 @@ export function projection(nodeIndex: number, selectorIndex: number = 0, attrs?:
+export function addToViewTree<T extends LView|LContainer>(lView: LView, lViewOrLContainer: T): T {
+  // TODO(benlesh/misko): This implementation is incorrect, because it always adds the LContainer to
+  // the end of the queue, which means if the developer asks for the LContainers out of order, the
+  // change detection will run out of order.
```
Maybe add clarification to comment about what "asks for" means?  Do you mean `injector.get(ViewContainerRef)`?  (Also is there a Jira ticket for this?)

kara replied:
^ Still needs resolution

kara commented on packages/core/src/render3/interfaces/view.ts:
```diff
@@ -85,7 +85,7 @@ export interface LView extends Array<any> {
    * destroy embedded views.
    */
-  [PARENT]: LView|null;
+  [PARENT]: LView|LContainer|null;
```
Can you update the docs comment above to include containers?

kara replied:
^ Still needs resolution

kara commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -7,28 +7,32 @@
-    const containerHostIndex = embeddedView[CONTAINER_INDEX];
-    return containerHostIndex > -1 ? embeddedView[PARENT] ![containerHostIndex] : null;
+    const container = embeddedView[PARENT];
+    return isLContainer(container) ? container : null;
```
The parent of a dynamic view isn't a container when it hasn't been attached yet, right?Think this needs a comment since it's not immediately obvious.

kara commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -7,28 +7,32 @@
     // This is a inline view node (e.g. embeddedViewStart)
-    return embeddedView[PARENT] ![tNode.parent !.index] as LContainer;
+    const parent = embeddedView[PARENT] !;
+    return isLContainer(parent) ? parent : parent[tNode.parent !.index] as LContainer;
```
When is the parent of an inline view not the container? It seems to me like this should be:

```ts
return embeddedView[PARENT];
```

kara commented on packages/core/src/render3/util.ts:
```diff
@@ -9,15 +9,33 @@
+ * Returns true if the value is an {@link LView}
+ * @param value the value to check
+ */
+export function isLView(value: any): value is LView {
```
I'm surprised we didn't have this yet ! Good add 👍

kara commented on packages/core/src/render3/util.ts:
```diff
@@ -146,7 +166,7 @@ export function getRootView(target: LView | {}): LView {
   let lView = Array.isArray(target) ? (target as LView) : readPatchedLView(target) !;
   while (lView && !(lView[FLAGS] & LViewFlags.IsRoot)) {
-    lView = lView[PARENT] !;
+    lView = lView[PARENT] !as LView;
```
We may walk up components that are inside embedded views

```suggestion
    lView = getLViewParent(lView) !;
```

Also, this looks almost identical to `getRootView` in `discovery_utils` (which you updated above as well). We should remove one of these as part of the cleanup.

benlesh replied:
Removed the other implementation of `getRootView`, and also a duplicate of `getRootContext`. :man_shrugging:

kara reviewed (changes requested):
Nice work, like how much more intuitive the data structures are.

benlesh committed:
fix(ivy): assertDomNode will now work properly in a worker

kara commented on packages/core/src/render3/discovery_utils.ts:
```diff
@@ -96,28 +96,18 @@ export function getContext<T = {}>(element: Element): T|null {
     // As long as lView[HOST] is null we know we are part of sub-template such as `*ngIf`
-    lView = lView[PARENT] !;
+    if (parent) {
+      lView = parent;
```
If `parent` is falsy, does this mean that `lView` never gets re-assigned? Wouldn't this create an infinite loop? 🤔

benlesh replied:
If the `parent` is falsy, it wouldn't pass the `while` loop's test condition.

kara replied:
So why do we need to check for it in line 104 then?

benlesh replied:
The if statement was redundant. Removed.

kara commented on packages/core/src/render3/interfaces/view.ts:
```diff
@@ -161,13 +162,25 @@ export interface LView extends Array<any> {
+   * Necessary to store this so views can traverse through their nested views
+   * to remove listeners and call onDestroy callbacks.
+   *
+   * For embedded views, we store the index of an LContainer's host rather than the first
```
^ Still needs resolution

benlesh commented:
@kara The TODO in `addViewToTree` is actually addressed in another PR that's in-progress (for FW-869). So that will not be resolved in this PR.

kara commented:
@benlesh To clarify, [the feedback that still needed resolution](https://github.com/angular/angular/pull/28382#discussion_r254950685) was that we need a better comment. I don't see why improving the comment needs to wait until the next PR.

> Maybe add clarification to comment about what "asks for" means? Do you mean `injector.get(ViewContainerRef)`? (Also is there a Jira ticket for this?)

kara commented on packages/core/src/render3/interfaces/view.ts:
```diff
@@ -79,14 +79,18 @@ export interface LView extends Array<any> {
    * views as it is recursively rendering templates.
    *
    * This is the "insertion" view for embedded views. This allows us to properly
    * destroy embedded views.
```
Looks like this line should be removed since embedded views will not have the insertion view as a parent anymore.

kara reviewed (approved):
LGTM, aside from the outdated lines

benlesh commented:
Caretaker: CI has been flakey for this one. Yarn refusing to update etc.

IgorMinar commented:
google3 presumit fails... please update and remove the "cleanup" label when ready

benlesh committed:
refactor(ivy): split util functions into different files

Google3 detected circular references here, so splitting up this rather hodge-podge list of functions into slightly better organizational units.

benlesh commented:
[new presubmit](https://test.corp.google.com/ui#id=OCL:232328029:BASE:235071797:1550788745754:83c6b2d9)

AndrewKushnir commented on packages/core/src/debug/debug_node.ts:
```diff
@@ -7,13 +7,14 @@
 import {getStylingContext} from '../render3/styling/util';
-import {INTERPOLATION_DELIMITER, isPropMetadataString, renderStringify} from '../render3/util';
+import {getComponent, getContext, getInjectionTokens, getInjector, getListeners, getLocalRefs, isBrowserEvents, loadLContext, loadLContextFromNode} from '../render3/util/discovery_utils';
+import {INTERPOLATION_DELIMITER, isPropMetadataString, renderStringify} from '../render3/util/misc_utils';
```
nit: I'm wondering if we could drop the `_utils` postfix (since it's already present in a path)?
`../render3/util/misc_utils` -> `../render3/util/misc`

benlesh replied:
when the file is open in your IDE, it makes it easier to find... in particular `interfaces/view.ts` vs `view.ts` vs `utils/view.ts`... in many IDEs that would just be `view.ts` `view.ts` `views.ts`.

AndrewKushnir replied:
Ok, thanks.

kara replied:
^ Yeah, I like that the `utils` is part of the file name.  `misc.ts` is pretty general on its own.

IgorMinar commented:
this pr still fails on g3 presubmit. adding "cleanup" label.

root cause:
```
failed to load main  Error: Cyclical dependency between files:
/build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/render3/util/view_utils.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/public_api.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/core.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/metadata.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/metadata/directives.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/render3/jit/directive.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/render3/util/misc_utils.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/render3/util/view_traversal_utils.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/render3/assert.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/render3/util/view_utils.ts
```

benlesh commented:
@IgorMinar that's an old presubmit, it's passing now: 

https://test.corp.google.com/ui#id=OCL:232328029:BASE:235220976:1550861966438:b7aecde3
